web/html/src/manager/state/highstate-summary.tsx

```diff
@@ -0,0 +1,137 @@
+import * as React from "react";
+import { useState, useEffect } from "react";
+import Network from "utils/network";
+import { Column } from "components/table/Column";
+import { Table } from "components/table/Table";
+import { AsyncButton } from "components/buttons";
+
+interface StateSource {
+  id: number;
+  name: string;
+  type: "STATE" | "CONFIG" | "FORMULA" | "INTERNAL";
+  sourceId: number;
+  sourceName: string;
+  sourceType: "SYSTEM" | "GROUP" | "ORG";
+}
+
+const typeMap = {
+  "STATE": t("State channel"),
+  "CONFIG": t("Config channel"),
+  "FORMULA": t("Formula"),
+  "INTERNAL": window._IS_UYUNI ? "Uyuni " : "SUSE Manager " + t("internal")
+};
+
+export default function HighstateSummary({ minionId }) {
+  const [summary, setSummary] = useState<StateSource[]>([]);
+  const [isLoading, setLoading] = useState(false);
+
+  useEffect(() => {
+    setLoading(true);
+    Network.get(`/rhn/manager/api/states/summary?sid=${minionId}`).promise
+      .then(setSummary)
+      .then(() => setLoading(false));
+  }, [minionId]);
+
+  if (isLoading) {
+    return (
+      <div className="row">
+        <span>{t("Retrieving highstate summary...")}</span>
+      </div>
+    );
+  }
+
+  return (
+    <>
+      <Table identifier={state => state.state} data={summary} initialItemsPerPage={0}>
+        <Column header={t("State")} columnKey="state" cell={source => <State minionId={minionId} state={source} />} />
+        <Column header={t("Type")} columnKey="type" cell={source => typeMap[source.type]} />
+        <Column header={t("Inherited From")} columnKey="source" cell={source => <Source source={source} />} />
+      </Table>
+      <HighstateOutput minionId={minionId} />
+    </>
+  );
+}
+
+function HighstateOutput({ minionId }) {
+  const [highstate, setHighstate] = useState("");
+
+  function requestHighstate(id: number) {
+    return Network.get(`/rhn/manager/api/states/highstate?sid=${id}`).promise;
+  }
+
+  if (!highstate) {
+    return (
+      <AsyncButton
+        icon="fa-file-text-o"
+        text={t("Show full highstate output")}
+        action={() => requestHighstate(minionId).then(setHighstate)}
+        defaultType="btn-link" />
+    );
+  }
+
+  return (
+    <div className="row">
+      <pre>{highstate}</pre>
+    </div >
+  );
+}
+
+function State({ minionId, state }: { minionId: number, state: StateSource }) {
+  if (state.type === "STATE") {
+    return (
+      <>
```

Comment:
I'm not sure, but there might be a small issue here: there's multiple children but no keys, I would expect there to be a warning in development mode? But perhaps I misunderstand how the pieces fit together here.

Reply:
In any case, only one of the fragments is returned so it shouldn't be a problem. There are no "key" warnings in the dev console either.